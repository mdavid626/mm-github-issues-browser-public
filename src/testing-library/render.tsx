/* istanbul ignore file */
import { InitialEntry, Location } from '@remix-run/router';
import { Queries, queries } from '@testing-library/dom';
import {
  RenderHookOptions,
  RenderOptions,
  render,
  renderHook,
  RenderResult,
  RenderHookResult,
} from '@testing-library/react';
import React, { ReactElement } from 'react';
import { MemoryRouter, useLocation, useNavigate } from 'react-router-dom';
import { NavigateFunction } from 'react-router/dist/lib/hooks';

type SimpleRouter = {
  location?: Location;
  navigate?: NavigateFunction;
};

type RouterResult = {
  router: SimpleRouter;
};

export const renderWithRouter = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialEntries?: InitialEntry[]
): RenderResult & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  } as RenderResult & RouterResult;
};

export const renderHookWithRouter = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialEntries?: InitialEntry[]
): RenderHookResult<Result, Props> & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  };
};
