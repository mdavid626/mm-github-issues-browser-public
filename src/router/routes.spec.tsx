import { cleanup, screen } from '@testing-library/react';
import React from 'react';
import AboutPage from '../pages/about-page/about-page';
import IssuePage from '../pages/issue-page/issue-page';
import IssuesPage from '../pages/issues-page/issues-page';
import { renderWithRouter } from '../testing-library/render';
import Routes from './routes';

jest.mock('../pages/issues-page/issues-page');
jest.mock('../pages/issue-page/issue-page');
jest.mock('../pages/about-page/about-page');

describe('routes', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error').mockReturnValue(undefined);
    (IssuesPage as jest.Mock).mockReturnValue(<div>issues-page</div>);
    (IssuePage as jest.Mock).mockReturnValue(<div>issue-page</div>);
    (AboutPage as jest.Mock).mockReturnValue(<div>about-page</div>);
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  const pages: string[] = ['/', '/issue/1', '/about'];
  it.each(pages)('should render "%s" page', (page) => {
    const { asFragment } = renderWithRouter(<Routes />, undefined, [page]);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render default page when unknown route', () => {
    renderWithRouter(<Routes />, undefined, ['/unknown']);
    expect(screen.getByText('issues-page')).toBeVisible();
  });

  it('should throw error on error route', () => {
    expect(() => renderWithRouter(<Routes />, undefined, ['/error'])).toThrow(
      'test error'
    );
  });
});
