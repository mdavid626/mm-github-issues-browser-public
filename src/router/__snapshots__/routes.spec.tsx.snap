// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes should render "/" page 1`] = `
<DocumentFragment>
  <div>
    issues-page
  </div>
</DocumentFragment>
`;

exports[`routes should render "/about" page 1`] = `
<DocumentFragment>
  <div>
    about-page
  </div>
</DocumentFragment>
`;

exports[`routes should render "/issue/1" page 1`] = `
<DocumentFragment>
  <div>
    issue-page
  </div>
</DocumentFragment>
`;
